src/Diff.hs

```diff
@@ -0,0 +1,606 @@
+{-# LANGUAGE NamedFieldPuns             #-}
+{-# LANGUAGE OverloadedStrings          #-}
+{-# LANGUAGE RecordWildCards            #-}
+
+module Diff where
+
+import Control.Monad (forM)
+import Control.Monad.IO.Class (MonadIO, liftIO)
+import Control.Monad.Reader (MonadReader, ReaderT, ask)
+import Control.Monad.State (MonadState, StateT, get, put)
+import Data.Attoparsec.Text (IResult(..))
+import Data.List.NonEmpty (NonEmpty(..))
+import Data.Map (Map)
+import Data.Maybe (catMaybes)
+import Data.Monoid ((<>))
+import Data.Set (Set)
+import Data.Text (Text)
+import Data.Vector (Vector)
+import Nix.Derivation (Derivation, DerivationOutput)
+
+import qualified Control.Monad.Reader
+import qualified Data.Attoparsec.Text
+import qualified Data.Char            as Char
+import qualified Data.List.NonEmpty
+import qualified Data.Map
+import qualified Data.Set
+import qualified Data.String          as String
+import qualified Data.Text            as Text
+import qualified Data.Vector
+import qualified Nix.Derivation
+import qualified Patience
+import qualified System.Directory     as Directory
+import qualified System.FilePath      as FilePath
+import qualified System.Process       as Process
+
+#if MIN_VERSION_base(4,9,0)
+import Control.Monad.Fail (MonadFail)
+import qualified Data.ByteString
+import qualified Data.Text.Encoding
+import qualified Data.Text.Encoding.Error
+#endif
+
+
+newtype Status = Status { visited :: Set Diffed }
+
+data Diffed = Diffed
+    { leftDerivation  :: FilePath
+    , leftOutput      :: Set Text
+    , rightDerivation :: FilePath
+    , rightOutput     :: Set Text
+    } deriving (Eq, Ord)
+
+newtype Diff a = Diff { unDiff :: ReaderT DiffContext (StateT Status IO) a }
+    deriving
+    ( Functor
+    , Applicative
+    , Monad
+    , MonadReader DiffContext
+    , MonadState Status
+    , MonadIO
+#if MIN_VERSION_base(4,9,0)
+    , MonadFail
+#endif
+    )
+
+data DiffContext = DiffContext
+  { orientation :: Orientation
+  , environment :: Bool
+  }
+
+data Orientation = Character | Word | Line
+
+data Changed a = Changed { before :: a, now :: a }
+
+type OutputHash = Text
+
+type Platform = Text
+
+type Builder = Text
+
+type Argument = Text
+
+data DerivationDiff
+  = DerivationsAreTheSame
+  | AlreadyCompared
+  | NamesDontMatch   { outputStructure :: Changed (FilePath, Set Text)}
+  | OutputsDontMatch { outputStructure :: Changed (FilePath, Set Text)}
+  | DerivationDiff
+      { outputStructure :: Changed (FilePath, Set Text)
+      , outputsDiff     :: OutputsDiff
+      , platformDiff    :: Maybe (Changed Platform)
+        -- ^ Will be Nothing, if Platform does not change
+      , builderDiff     :: Maybe (Changed Builder)
+        -- ^ Will be Nothing, if Builder does not change
+      , argumentsDiff   :: Maybe [Patience.Item Argument]
+        -- ^ Will be Nothing, if arguments are equal
+      , sourcesDiff     :: SourcesDiff
+      , inputExtraNames :: Maybe (Changed (Set Text))
```

Comment:
Cannot `inputExtraNames` and `inputsDiff` be unified, so that when one derivation has an extra input, it has the corresponding item in `inputsDiff` list?

Reply:
No, in `inputExtraNames` there are inputs, with names that exist in one derivation and doesn't in other one. And `inputsDiff` is comparison of inputs with the same name. But I can move these fields from the `DerivationDiff`'s body, as I did for `SourcesDiff`and `EnvironmentDiff`.

Reply:
Hmm, Do these two situations (when input exists only in one derivation and when input exists in both) render differently?

Reply:
Yes, inputExtraNames and inputsDiff would be rendered separately.